### DISCUSSION TITLE
Unexpected CORS  / Web Origins results

### DISCUSSION THREAD

**Discussion by meussiebeaucoup** [keycloak/keycloak#30797] [Q&A]
Hi,

This was tested on several versions, 23.x up to and including 25.x:

I'm running into some unexpected results when setting the `Web Origins` configuration for a client. I've setup the field so that it contains the following value:

![image](https://github.com/keycloak/keycloak/assets/68510105/e6b9002e-6862-4db0-be09-4ca50e0ef03f)

My assumption is that every request with a `Origin` header containing values other than `https://good-origin/` would be rejected, but that is not the case.

The preflight response is as follows:

```
OPTIONS http://localhost:8080/auth/realms/Test/protocol/openid-connect/token
Content-Type: application/x-www-form-urlencoded
Origin: http://google.com
Connection: Keep-Alive
User-Agent: Apache-HttpClient/4.5.14 (Java/17.0.9)
Accept-Encoding: br,deflate,gzip,x-gzip
```

This results in the following response:

```
HTTP/1.1 200 OK
Access-Control-Allow-Origin: http://google.com
Access-Control-Allow-Methods: POST, OPTIONS
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Origin, Accept, X-Requested-With, Content-Type, Access-Control-Request-Method, Access-Control-Request-Headers, DPoP, Authorization
Access-Control-Max-Age: 3600
Content-Security-Policy: frame-src 'self'; frame-ancestors 'self'; object-src 'none';
Referrer-Policy: no-referrer
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Robots-Tag: none
X-XSS-Protection: 1; mode=block
content-length: 0
```

I would assume that the `Access-Control-Allow-Origin` field would not contain http://google.com, since this is the origin from the request and not the Web Origin.

Furthermore, if I perform the actual POST request to the token endpoint, in this case a direct access grant:

```
POST http://localhost:8080/auth/realms/Test/protocol/openid-connect/token
Content-Type: application/x-www-form-urlencoded
Origin: http://google.com

client_id = portal &
username = test &
password = test &
grant_type = password
```

Again I assumed this would result in an error, since the origins don't match, but to my suprise I get a token response:

```
{
  "access_token": "eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJaQVo5UEI5UEdHMDR4N3hiZ0YtV0twM2RwNThtY09lM3Q3REM3SlVZcWtBIn0....
  "not-before-policy": 0,
  "session_state": "144882f6-8074-4d1d-b7ef-d9b5c7b865e3",
  "scope": "profile email"
}
```

Am I missing something, is this working as intended? Is there a way I can prevent token retrieval from origins other than the ones specified within Keycloak, or should this be done outside Keycloak?

**Comment by nesika** [comment]
I'm facing the same issue. I have set the Web Origin with one url (https://www.google.com) which should be allowed when hitting the realm api of keycloak. But when I set the header origin with some other url(evil.com) in the postman application, the result is Access-Control-Allow-Origin is set with evil.com. In this case I expect to Access-Control-Allow-Origin header shouldn't present and the CORS error should be displayed on the browser console as the mentioned origin is not listed in the Web Origins. 

Even I tried adding the url in the 'Valid redirect URIs' in the keycloak client settings and + symbol in the Web Origin to fix the issue. Can someone please help here how we can restrict the various origins other than the urls which are mentioned in the 'Web Origins' in the keycloak console.

**Comment by danielFesenmeyer** [comment]
I had to look into this topic some time ago, that's how I understand it:

I think the preflight response is not relevant in this case, because a preflight request DOES NOT send the Authorization header.
The "problem" now is that Keycloak does not have the Authorization header, which it requires to read the bearer token with the client ID. Without client ID, Keycloak cannot figure out which origins are allowed and therefore has to allow any preflight request. Otherwise it would not be possible to serve ANY request.

Therefore, only subsequent (non-preflight) requests are relevant in this scenario. These requests contain Authorization data, which is checked by Keycloak, and invalid web origins should be rejected.

**Reply by danielFesenmeyer** [reply]
I just see my comment does not explain why it's possible to _create_ a token from an invalid web origin.

But actually, I think it's more important not being able to _use_ a token from an invalid web origin. And that's what I think is ensured by Keycloak.
